File: year2021/day05/python/solution.py
def main():
    print(part_1())
    print(part_2())


def part_1():
    seen, duplicated = set(), set()
    for row in input:
        if not (row[0][0] != row[1][0] and row[0][1] != row[1][1]):
            seen, duplicated = check_row(iterate(row), seen, duplicated)

    return len(duplicated)


def part_2():
    seen, duplicated = set(), set()
    for row in input:
        seen, duplicated = check_row(iterate(row), seen, duplicated)

    return len(duplicated)


def setup(path):
    global input
    
    with open(path, 'r') as file:
        input = [[[int(number) for number in side.split(',')] for side in row.split(' -> ')] for row in file.read().split('\n')]


def check_row(iterator, seen, duplicated):
    iteration = next(iterator, None)
    while iteration is not None:
        x, y = iteration
        coordinate = f"{x},{y}"

        if coordinate in seen:
            duplicated.add(coordinate)

        seen.add(coordinate)
        iteration = next(iterator, None)

    return seen, duplicated


def iterate(row):
    dx, dy = (1 if row[1][0] > row[0][0] else -1 if row[1][0] < row[0][0] else 0, 1 if row[1][1] > row[0][1] else -1 if row[1][1] < row[0][1] else 0)
    x, y = row[0][0], row[0][1]
    yield x, y

    while (x, y) != (row[1][0], row[1][1]):
        if x != row[1][0]:
            x += dx

        if y != row[1][1]:
            y += dy

        yield x, y


if __name__ == '__main__':
    setup('../puzzle.txt')
    main()

File: year2021/day05/java/Solution.java
import java.nio.file.*;
import java.util.*;
import java.util.stream.Collectors;

public class Solution {
  private static List<Coordinates> COORDINATES;

  public static void main(String[] args) {
    try {
      setup(Files.readAllLines(Paths.get("../puzzle.txt")));
      System.out.println(part1());
      System.out.println(part2());
    } catch(Exception e) {
      e.printStackTrace();
    }
  }

	public static int part1() {
    Map<String, Integer> field = new HashMap<>();
    COORDINATES.stream()
        .filter(command -> command.x1 == command.x2 || command.y1 == command.y2)
        .collect(Collectors.toList())
        .forEach(command -> calcField(command, field));
    return (int) field.entrySet().stream().filter(entry -> entry.getValue() >= 2).count();
  }

  public static int part2() {
    Map<String, Integer> field = new HashMap<>();
    COORDINATES.stream()
        .filter(command -> command.x1 == command.x2 || command.y1 == command.y2)
        .collect(Collectors.toList())
        .addAll(COORDINATES.stream().filter(command -> Math.abs(command.x1 - command.x2) == Math.abs(command.y1 - command.y2))
        .collect(Collectors.toList()));
    COORDINATES.forEach(command -> calcField(command, field));
    return (int) field.entrySet().stream().filter(entry -> entry.getValue() >= 2).count();
  }

  public static void setup(List<String> input) {
    COORDINATES = input.stream().map(Coordinates::createFromInput).collect(Collectors.toList());
  }

  private static void calcField(Coordinates coordinates, Map<String, Integer> field) {
    int startX = coordinates.x1;
    int startY = coordinates.y1;
    int endX = coordinates.x2;
    int endY = coordinates.y2;
    for (int i = 0; i <= Math.max(Math.abs(endX - startX), Math.abs(endY - startY)); i++) {
      field.merge((startX + i * Integer.compare(endX, startX)) + "," + (startY + i * Integer.compare(endY, startY)), 1, Integer::sum);
    }
  }

}

class Coordinates {
  public int x1, x2, y1, y2;

  public Coordinates(int x1, int x2, int y1, int y2) {
    this.x1 = x1;
    this.x2 = x2;
    this.y1 = y1;
    this.y2 = y2;
  }

  public static Coordinates createFromInput(String inputLine) {
    String[] splitted = inputLine.split("->");
    String[] from = splitted[0].trim().split(",");
    String[] to = splitted[1].trim().split(",");
    return new Coordinates(Integer.parseInt(from[0]), Integer.parseInt(to[0]), Integer.parseInt(from[1]), Integer.parseInt(to[1]));
  }

  @Override
  public String toString() {
    return "Command{" + x1 + "," + y1 + " -> " + x2 + "," + y2 + '}';
  }
}